pkg/model/api/v1/datasource.go

```diff
@@ -16,117 +16,142 @@ package v1
 import (
 	"encoding/json"
 	"fmt"
-	"net/url"
+
+	"github.com/perses/perses/pkg/model/api/v1/datasource"
+	"gopkg.in/yaml.v2"
 )
 
-func GenerateDatasourceID(name string) string {
-	return fmt.Sprintf("/datasources/%s", name)
+func GenerateGlobalDatasourceID(name string) string {
+	return fmt.Sprintf("/globaldatasources/%s", name)
 }
 
-type DatasourceSpec struct {
-	URL *url.URL `json:"url" yaml:"url"`
+func GenerateDatasourceID(project string, name string) string {
+	return generateProjectResourceID("datasources", project, name)
 }
 
-type tmpDatasourceSpec struct {
-	URL string `json:"url" yaml:"url"`
+type DatasourceSpec interface {
+	GetKind() datasource.Kind
 }
 
-func (d *DatasourceSpec) MarshalJSON() ([]byte, error) {
-	urlAsString := ""
-	if d.URL != nil {
-		urlAsString = d.URL.String()
-	}
-	tmp := &tmpDatasourceSpec{
-		URL: urlAsString,
+func unmarshalDatasourceSpec(spec map[string]interface{}, staticMarshal func(interface{}) ([]byte, error), staticUnmarshal func([]byte, interface{}) error) (DatasourceSpec, error) {
+	if specKind, ok := spec["kind"]; !ok {
+		return nil, fmt.Errorf("attribute 'kind' not found in 'datasource.spec'")
+	} else {
```

Comment:
The `else` isn't needed here, as you return early in the first branch anyway, so you can keep things less nested / indented.

Reply:
Ah use it to access `specKind` later, ok.